Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
 #else
     UNUSED_PARAM(didWrite);
 #endif
+}
+
+void NetworkStorageManager::spaceGrantedForOrigin(const WebCore::ClientOrigin& origin, uint64_t spaceRequested)
+{
+    assertIsCurrent(workQueue());
+
+    updateLastModificationTimeForOrigin(origin);
+    if (!m_totalQuotaRatio)
+        return;
+
+    if (!m_totalQuota) {
+        std::optional<uint64_t> volumeCapacity;
+        if (m_volumeCapacityOverride)
+            volumeCapacity = m_volumeCapacityOverride;
+        else if (auto capacity = FileSystem::volumeCapacity(m_path))
+            volumeCapacity = WTF::roundUpToMultipleOf(defaultVolumeCapacityUnit, *capacity);
+        if (volumeCapacity)
+            m_totalQuota = *m_totalQuotaRatio * *volumeCapacity;
+        else
+            return;
+    }
 
-    if (m_quotaManager)
-        m_quotaManager->originVisited(origin);
+    if (m_totalUsage)
+        m_totalUsage = *m_totalUsage + spaceRequested;
+
+    if (!m_totalUsage || *m_totalUsage > *m_totalQuota)
+        schedulePerformEviction();
 }
 
-NetworkQuotaManager& NetworkStorageManager::quotaManager()
+void NetworkStorageManager::schedulePerformEviction()
 {
-    if (!m_quotaManager) {
-        Vector<std::pair<WebCore::ClientOrigin, WallTime>> origins;
-        forEachOriginDirectory([&](auto directory) {
-            auto originFile = originFilePath(directory);
-            auto modificationTime = FileSystem::fileModificationTime(originFile);
-            if (modificationTime) {
-                if (auto origin = readOriginFromFile(originFile))
-                    origins.append({ *origin, *modificationTime });
-            }
-        });
+    assertIsCurrent(workQueue());
+
+    if (m_isEvictionScheduled)
+        return;
+
+    m_isEvictionScheduled = true;
+    getDomainsWithLastAccessedTimeForEviction();
+}
+
+void NetworkStorageManager::getDomainsWithLastAccessedTimeForEviction()
+{
+    assertIsCurrent(workQueue());
+
+    RunLoop::main().dispatch([this, weakThis = ThreadSafeWeakPtr { *this }]() {
+        auto strongThis = weakThis.get();
+        if (!strongThis || m_closed || !m_process)
+            return;
 
-        std::sort(origins.begin(), origins.end(), [&](const auto& a, const auto& b) {
-            return a.second < b.second;
+        m_process->registrableDomainsWithLastAccessedTime(m_sessionID, [this, weakThis = ThreadSafeWeakPtr { *this }](auto result) mutable {
```

Comment:
> I think it is important we continue deleting from top level origins, it removes the possibility for third-party iframes to try communicating cross domain by observing deletion of third party data in case usage is close to quota.
> First party can still do this but this requires top level navigation which is potentially more costly.

hmm I don't quite get this comment; are you saying we might not delete by top-level origin with this patch?
`performEviction` takes a map keyed by top level origin, so it's supposed to delete by top level origin.